design-documents/buildattr64-rationale.rst

```diff
@@ -0,0 +1,540 @@
+
+An example of a feature that is not amenable to runtime checking is
+use of pointer authentication instructions outside of the hint space.
+These could be present in all non-leaf functions so testing and
+providing alternative implementations would be prohibitively
+expensive.
+
+An example of a feature that must enforce a property across a
+loadable-unit is branch target identification (BTI). When this feature
+is enabled for a loadable-unit all indirect branch targets in the
+loadable-unit must have a BTI compatible landing pad.
+
+Design Goals
+============
+
+The goals of the AArch64 build attributes specification are to:
+
+- Define the encoding for AArch64 build attributes in relocatable
+  object files.
+
+- Define the relationship between build attributes and the existing
+  GNU program properties.
+
+- Reuse as much of what worked well from the AArch32 Build Attributes
+  in (ADDENDA32_).
+
+- Make it easier for a build attributes consumer to skip a subsection
+  or attribute that it does not understand without giving a warning
+  message.
+
+- Separate architectural requirements from software ABI requirements.
+
+Build Attributes or GNU properties?
+===================================
+
+There are many ways that attributes about a program can be represented,
+both in the abstract and in the concrete way that they are encoded in
+an ELF file. GNU properties the most likely alternative as they
+are already used in (AAELF64_) for ``GNU_PROPERTY_AARCH64_FEATURE_1_BTI``
+and are extensively used in the (X86_64PSABI_). GNU properties could be
+extended for all relocatable object marking instead of Build Attributes.
+
+From an encoding perspective the GNU property format as defined by
+(LINUX_ABI_) permits any data in the program property array. There could
+be a ``GNU_PROPERTY_AARCH64_ATTRIBUTES`` property which contain the
+same information as (BUILDATTR64_) defines for the ``SHT_AARCH64_ATTRIBUTES``
+section. This would essentially be using GNU properties as a build attributes
+wrapper.
+
+From a modelling perspective GNU properties are present in both
+relocatable objects and loadable-units which influences the design of
+the properties with a simple combination rule per property for the
+static linker to transfer the relocatable-object properties to the
+loadable-unit. For example:
+
+- ``GNU_PROPERTY_STACK_SIZE`` has the static linker copy the maximum
+  value of all the relocatable object inputs
+  ``GNU_PROPERTY_STACK_SIZE`` to the output property.
+
+- ``GNU_PROPERTY_AARCH64_FEATURE_1_AND`` transfers the ``AND`` of all
+  the feature bits from the relocatable object inputs to the output
+  property. In practice an additional input for some the feature bits
+  has come from the static linker, making the implementation less
+  generic than it could be.
+
+The major differences between build attributes and GNU properties are:
+
+- Build attributes are relocatable object only and have an encoding
```

Comment:
This sentence appears to contradict with "Encoding of meta data in a loadable-unit" in `buildattr64.rst`. If the recommended practice is for loaders not to inspect the platform-specific build attributes, mention it in `buildattr64.rst`?

Reply:
In `buildattr64.rst` there is:
```
Encoding of meta data in a loadable-unit
----------------------------------------

Build attributes are only defined for relocatable objects.  The
encoding of metadata in a loadable-unit (executable or shared-library)
is platform specific.  For example a platform may choose to use GNU
program properties as defined in (LINUX_ABI_). This document specifies
the meaning of build attributes for relocatable objects only.  The
presence of ``.ARM.attributes`` sections in a loadable-unit is
permitted, but the interpretation of the contents of the section is
outside the scope of this document.
```

I think that is consistent with `Build Attributes are relocatable only`.

Arm has traditionally taken the approach that the executable and shared-library are mostly "owned" by the platform holder, be that Android or the Linux community and the AArch64 ABI. I've written https://github.com/ARM-software/abi-aa/pull/229 which is a rationale for using GNU properties in sysvabi64.rst files which we equate to Linux/BSD.

Hope that makes sense. Any suggestions for improvement are welcome though.